### PR TITLE
C#: Avoid inefficient DenseTensor ctor in ToTensor extensions

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 

Constructing DenseTensor using System.Array is inefficient as it involves boxing/unboxing for every value. We have an internal constructor for DenseTensor that takes System.Array that is used by the `ToTensor` methods in ArrayTensorExtensions. 

Update the extension methods to only use the inefficient ctor if they cannot flatten the input array directly (no boxing/unboxing if it can) and construct a DenseTensor with a `Memory<T>` around that copy.

Add extension method for 4D array. Assuming that is somewhat common if you have NCHW input.

I did a basic perf test of creating a DenseTensor and populating it via operator[] with indices and doing so is roughly equivalent to directly populating a 1D array, so that path to creating a DenseTensor is okay (and our existing samples which setup input that way are fine).

Misc: Fix some long lines 

**Motivation and Context**
#1652 and recurring issues with unexpected performance when constructing a DenseTensor from an existing multidimensional C# array.

### CODE REVIEW

File: csharp/src/Microsoft.ML.OnnxRuntime/Tensors/ArrayTensorExtensions.shared.cs [hunk 1]
```diff
@@ -27,7 +27,12 @@ public static class ArrayTensorExtensions
         /// <returns>A 1-dimensional DenseTensor&lt;T&gt; with the same length and content as <paramref name="array"/>.</returns>
         public static DenseTensor<T> ToTensor<T>(this T[] array)
         {
-            return new DenseTensor<T>(array);
```
Comment (gwang-msft) [inline csharp/src/Microsoft.ML.OnnxRuntime/Tensors/ArrayTensorExtensions.shared.cs:30]: Just curious, can all these copy logics here be part of the DenseTensor ctor?

Comment (skottmckay) [reply csharp/src/Microsoft.ML.OnnxRuntime/Tensors/ArrayTensorExtensions.shared.cs:30]: I believe DenseTensor was originally copied from a .net library so doing it this way preserved that interface and would make future updates cleaner.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.Common/Tensors/ArrayTensorExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        [Fact]
+        public void ConstructFrom1D()
+        {
+            var array = new int[] { 1, 2, 3, 4 };
+            var tensor = array.ToTensor();
+
+            var expectedDims = new int[] { 4 };
+            Assert.Equal(tensor.Length, array.Length);
+            Assert.Equal(expectedDims, tensor.Dimensions.ToArray());
```
Comment (gwang-msft) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests.Common/Tensors/ArrayTensorExtensionsTests.cs:21]: nit, verify the content of the tensor?

**Review by gwang-msft — APPROVED** [review]

**Review by yuslepukhin — DISMISSED** [review]
:shipit:

**Review by gwang-msft — APPROVED** [review]

**Pull request merged**